src/functional/iterator/embedded.py

```diff
@@ -318,16 +324,25 @@ def execute_shift(
     pos: Position, tag: Tag, index: IntIndex, *, offset_provider: OffsetProvider
 ) -> MaybePosition:
     assert pos is not None
-    if tag in pos and pos[tag] is None:  # sparse field with offset as neighbor dimension
-        new_pos = pos | {tag: index}
+    if isinstance(tag, SparseOffset):
+        new_pos = deepcopy(pos)
+
+        cur_pos = new_pos[tag]
+        assert isinstance(cur_pos, list)
+        for i, p in reversed(
+            list(enumerate(cur_pos))
+        ):  # first shift applies to the last sparse dimensions of that axis type
+            if p is None:
+                cur_pos[i] = index
+                break
         return new_pos
```

Comment:
```suggestion
        assert isinstance(pos[tag], list)
        new_entry = list(pos[tag])
        # assert None in new_entry  # Fail instead of ignore too many partial shifts
        for i, p in reversed(enumerate(new_entry)):
            # first shift applies to the last sparse dimensions of that axis type
            if p is None:
                new_entry[i] = index
                break
        return pos | {tag: new_entry}
```

Reply:
took it with smaller changes, but couldn't make mypy happy...